File: src/main/java/club/xiaojiawei/status/War.java
package club.xiaojiawei.status;

import club.xiaojiawei.bean.Card;
import club.xiaojiawei.bean.log.ExtraEntity;
import club.xiaojiawei.bean.Player;
import club.xiaojiawei.bean.log.TagChangeEntity;
import club.xiaojiawei.bean.area.Area;
import club.xiaojiawei.controller.javafx.MainController;
import club.xiaojiawei.enums.StepEnum;
import club.xiaojiawei.enums.WarPhaseEnum;
import club.xiaojiawei.enums.ZoneEnum;
import javafx.beans.property.SimpleIntegerProperty;
import lombok.Getter;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;

import java.util.Objects;
import java.util.concurrent.atomic.AtomicInteger;

import static club.xiaojiawei.data.ScriptStaticData.CARD_AREA_MAP;
import static club.xiaojiawei.enums.WarPhaseEnum.FILL_DECK_PHASE;

/**
 * 游戏对局状态
 * @author 肖嘉威
 * @date 2022/11/25 20:57
 */
@Slf4j
public class War {

    @Getter
    private volatile static Player currentPlayer;
    @Getter
    private static String firstPlayerGameId;
    @Getter
    @Setter
    private volatile static WarPhaseEnum currentPhase = FILL_DECK_PHASE;
    @Getter
    private volatile static StepEnum currentTurnStep;
    @Getter
    @Setter
    private volatile static Player me;
    @Getter
    @Setter
    private volatile static Player rival;
    @Getter
    private volatile static Player player1;
    @Getter
    private volatile static Player player2;
    @Getter
    @Setter
    private volatile static int warTurn;
    @Getter
    @Setter
    private volatile static String won;
    @Getter
    @Setter
    private volatile static String lost;
    @Getter
    @Setter
    private volatile static String conceded;
    @Setter
    @Getter
    private volatile static long startTime;
    @Getter
    @Setter
    private volatile static long endTime;
    @Getter
    @Setter
    private volatile static boolean myTurn;

    public final static SimpleIntegerProperty WAR_COUNT = new SimpleIntegerProperty();
    public final static AtomicInteger WIN_COUNT = new AtomicInteger();
    /**
     * 单位：min
     */
    public final static AtomicInteger GAME_TIME = new AtomicInteger();
    public final static AtomicInteger EXP = new AtomicInteger();

    public static void reset(){
        firstPlayerGameId = null;
        currentPhase = FILL_DECK_PHASE;
        currentTurnStep = null;
        currentPlayer = me = rival = null;
        player1 = new Player("1");
        player2 = new Player("2");
        warTurn = 0;
        won = lost = conceded = "";
        startTime = endTime = 0;
        CARD_AREA_MAP.clear();
        log.info("已重置游戏状态");
    }

    public static void setCurrentTurnStep(StepEnum currentTurnStep) {
        log.info((War.currentTurnStep = currentTurnStep).getComment());
    }

    public static synchronized void increaseWarCount(){
        boolean flag = false;
        if (War.getMe() != null){
            flag = printResult();
        }
        long time = (endTime - startTime) / 1000 / 60;
        log.info("本局游戏时长：" + time + "分钟");
        GAME_TIME.set((int) (time + GAME_TIME.get()));
        int winExp, lostExp;
        switch (MainController.getCurrentRunMode()){
            case STANDARD, WILD,CLASSIC, TWIST -> {
                winExp = 8;
                lostExp = 6;
            }
            case CASUAL, BACON->{
                winExp = 6;
                lostExp = 4;
            }
            default -> {
                winExp = 0;
                lostExp = 0;
            }
        }
        long earnExp = Math.min(time, 30) * (flag ? winExp : lostExp);
        log.info("本局游戏获得经验值：" + earnExp);
        EXP.set((int) (EXP.get() + earnExp));
        WAR_COUNT.set(WAR_COUNT.get() + 1);
    }

    private static boolean printResult(){
        boolean flag = false;
        if (Objects.equals(War.getWon(), War.getMe().getGameId())){
            War.WIN_COUNT.incrementAndGet();
            flag = true;
        }
        log.info("本局游戏胜者：" + won);
        log.info("本局游戏败者：" + lost);
        log.info("本局游戏投降者：" + conceded);
        return flag;
    }

    public static void setFirstPlayerGameId(String firstPlayerGameId) {
        log.info("先手玩家：" + (War.firstPlayerGameId = firstPlayerGameId));
    }

    public static void setCurrentPlayer(Player currentPlayer) {
        log.info((War.currentPlayer = currentPlayer).getGameId() + " 的回合");
    }

    public static Card exchangeAreaOfCard(ExtraEntity extraEntity){
        Area sourceArea = CARD_AREA_MAP.get(extraEntity.getEntityId());
        if (sourceArea == null){
            sourceArea = War.getPlayer(extraEntity.getPlayerId()).getArea(extraEntity.getZone());
        }
        Area targetArea = War.getPlayer(extraEntity.getPlayerId()).getArea(extraEntity.getExtraCard().getZone());
        Card card = sourceArea.removeByEntityId(extraEntity.getEntityId());
        card.updateByExtraEntity(extraEntity);
        targetArea.add(card, extraEntity.getExtraCard().getZonePos());
        return card;
    }
    public static void exchangeAreaOfCard(TagChangeEntity tagChangeEntity){
        Area sourceArea = CARD_AREA_MAP.get(tagChangeEntity.getEntityId());
        Area targetArea = War.getPlayer(tagChangeEntity.getPlayerId()).getArea(ZoneEnum.valueOf(tagChangeEntity.getValue()));
        targetArea.add(sourceArea.removeByEntityId(tagChangeEntity.getEntityId()), 0);
    }

    public static Player getPlayer(String playerId){
        return player1.getPlayerId().equals(playerId)? player1 : player2;
    }
    public static Player getReversePlayer(String playerId){
        return player1.getPlayerId().equals(playerId)? player2 : player1;
    }


    public static Player getPlayerByArea(Area area){
        if (area == player1.getPlayArea()
                || area == player1.getHandArea()
                || area == player1.getDeckArea()
                || area == player1.getGraveyardArea()
                || area == player1.getRemovedfromgameArea()
                || area == player1.getSecretArea()
                || area == player1.getSetasideArea()
        ){
            return player1;
        }
        return player2;
    }
    public static Area getReverseArea(Area area){
        if (area == player1.getPlayArea()){
            return player2.getPlayArea();
        }else if (area == player1.getHandArea()){
            return player2.getHandArea();
        }else if (area == player1.getDeckArea()){
            return player2.getDeckArea();
        }else if (area == player1.getGraveyardArea()){
            return player2.getGraveyardArea();
        }else if (area == player1.getRemovedfromgameArea()){
            return player2.getRemovedfromgameArea();
        }else if (area == player1.getSecretArea()){
            return player2.getSecretArea();
        }else if (area == player1.getSetasideArea()){
            return player2.getSetasideArea();
        }else if (area == player2.getPlayArea()){
            return player1.getPlayArea();
        }else if (area == player2.getHandArea()){
            return player1.getHandArea();
        }else if (area == player2.getDeckArea()){
            return player1.getDeckArea();
        }else if (area == player2.getGraveyardArea()){
            return player1.getGraveyardArea();
        }else if (area == player2.getRemovedfromgameArea()){
            return player1.getRemovedfromgameArea();
        }else if (area == player2.getSecretArea()){
            return player1.getSecretArea();
        }else if (area == player2.getSetasideArea()){
            return player1.getSetasideArea();
        }
        return null;
    }
}

File: Hearthstone-Script-Card-SDK/src/main/kotlin/club/xiaojiawei/status/War.kt
package club.xiaojiawei.status

import club.xiaojiawei.bean.Card
import club.xiaojiawei.bean.Player
import club.xiaojiawei.bean.area.Area
import club.xiaojiawei.bean.safeRun
import club.xiaojiawei.config.log
import club.xiaojiawei.enums.RunModeEnum
import club.xiaojiawei.enums.StepEnum
import club.xiaojiawei.enums.WarPhaseEnum
import club.xiaojiawei.mapper.WarMapper
import club.xiaojiawei.status.War.Companion.UNKNOWN_WAR
import club.xiaojiawei.util.isTrue

/**
 * 游戏对局状态
 * @author 肖嘉威
 * @date 2022/11/25 20:57
 */

/**
 * 全局war
 */
val WAR: War = War(true)

class War(
    /**
     * 允许打印日志
     */
    val allowLog: Boolean = false
) : Cloneable {

    /**
     * 存储当前war中所有的[Card]
     * k:[club.xiaojiawei.bean.Entity.entityId]
     */
    val cardMap: MutableMap<String?, Card?> = HashMap()

    @Volatile
    var currentPlayer: Player = Player.UNKNOWN_PLAYER
        @Synchronized set(value) {
            field = value
            value.safeRun {
                allowLog.isTrue {
                    log.info { "${value.gameId} 的回合" }
                }
            }
        }

    @Volatile
    var firstPlayerGameId: String = ""
        @Synchronized set(value) {
            field = value
            (value.isNotEmpty() && allowLog).isTrue {
                log.info { "先手玩家：$value" }
            }
        }

    @Volatile
    var currentPhase = WarPhaseEnum.FILL_DECK

    @Volatile
    var currentTurnStep: StepEnum? = null
        @Synchronized set(value) {
            field = value
            field?.let {
                allowLog.isTrue {
                    log.info { it.comment }
                }
            }
        }

    @Volatile
    var me: Player = Player.UNKNOWN_PLAYER

    @Volatile
    var rival: Player = Player.UNKNOWN_PLAYER

    @Volatile
    var player1: Player = Player.UNKNOWN_PLAYER

    @Volatile
    var player2: Player = Player.UNKNOWN_PLAYER

    /**
     * 总回合数
     */
    @Volatile
    var warTurn = 0

    /**
     * 胜者
     */
    @Volatile
    var won: String = ""

    /**
     * 败者
     */
    @Volatile
    var lost: String = ""

    /**
     * 投降者
     */
    @Volatile
    var conceded: String = ""

    /**
     * 本局开始时间
     */
    @Volatile
    var startTime: Long = 0

    /**
     * 本局结束时间
     */
    @Volatile
    var endTime: Long = 0

    @Volatile
    var isMyTurn = false

    @Volatile
    var currentRunMode: RunModeEnum? = null

    /**
     * 所生成最大的entityId，entityId越大表明entity生成的时间越晚
     */
    @Volatile
    var maxEntityId: String? = null

    /**
     * 最大entityId自增1
     */
    fun incrementMaxEntityId(): String {
        val nextMaxEntityId = ((maxEntityId?.toInt() ?: 0) + 1).toString()
        this.maxEntityId = nextMaxEntityId
        return nextMaxEntityId
    }

    fun exchangePlayer() {
        val tempMe = me
        me = rival
        rival = tempMe
        val tempPlayer1 = player1
        player1 = player2
        player2 = tempPlayer1
    }

    /**
     * 添加卡牌，当新生成的[Card]需要添加至[Area]时推荐用此方法代替[Area.add]，这会缓存此[Card]
     */
    fun addCard(card: Card, area: Area, pos: Int? = null) {
        cardMap[card.entityId] = card
        if (pos == null) {
            area.add(card)
        } else {
            area.add(card, pos)
        }
    }

    public override fun clone(): War {
        return deepClone(false)
    }

    fun deepClone(allowLog: Boolean): War {
        val newWar = War(allowLog = allowLog)
        WarMapper.INSTANCE.update(this, newWar)
        val newMe = this.me.deepClone(newWar)
        val newRival = this.rival.deepClone(newWar)
        newWar.me = newMe
        newWar.rival = newRival
        if (this.me == this.player1) {
            newWar.player1 = newMe
            newWar.player2 = newRival
        } else {
            newWar.player1 = newRival
            newWar.player2 = newMe
        }
        if (this.me == this.currentPlayer) {
            newWar.currentPlayer = newMe
        } else {
            newWar.currentPlayer = newRival
        }
        return newWar
    }

    companion object {
        val UNKNOWN_WAR: War = War(false)
    }
}

fun War.safeRun(block: (War) -> Unit): War {
    if (isInValid()) {
        return this
    } else {
        block(this)
        return this
    }
}

fun War.isValid(): Boolean {
    return this !== UNKNOWN_WAR
}

fun War.isInValid(): Boolean {
    return this === UNKNOWN_WAR
}

